libs/text-splitters/langchain_text_splitters/nltk.py

```diff
@@ -9,23 +9,46 @@ class NLTKTextSplitter(TextSplitter):
     """Splitting text using NLTK package."""
 
     def __init__(
-        self, separator: str = "\n\n", language: str = "english", **kwargs: Any
+        self,
+        separator: str = "\n\n",
+        language: str = "english",
+        use_span_tokenize: bool = False,
+        **kwargs: Any,
     ) -> None:
         """Initialize the NLTK splitter."""
         super().__init__(**kwargs)
+        self._separator = separator
+        self._language = language
+        self._use_span_tokenize = use_span_tokenize
+        if self._use_span_tokenize and self._separator != "":
+            raise ValueError("When use_span_tokenize is True, separator should be ''")
         try:
-            from nltk.tokenize import sent_tokenize
+            if self._use_span_tokenize:
+                from nltk.tokenize import _get_punkt_tokenizer
```

Comment:
is this recommended by nltk? I don't love importing a `_` function. Is there a way this gets called by `sent_tokenize` or another public function under the hood that would be better?

Reply:
This seems like a hack to get this working in the short-term, when in reality we should probably have the `self._tokenizer` be of type `TokenizerI`, and this should just call `span_tokenize` under the hood for this arg. Would that make sense to you? 

https://github.com/nltk/nltk/blob/develop/nltk/tokenize/api.py#L20

Reply:
Without some of the associated cleanup I don't think this PR is mergable in its current state :(. Let me know if you're willing to take this on and in which timeframe. Otherwise will probably comment something along these lines on the issue and close this PR

Reply:
Hi @efriis,


> This seems like a hack to get this working in the short-term, when in reality we should probably have the self._tokenizer be of type TokenizerI, and this should just call span_tokenize under the hood for this arg. Would that make sense to you?


I totally agree with you... I've implemented in this way just to don't break the current `sent_tokenize` logic, but I'm happy to refactor it. In this way, we'll integrate a cleaner code.

---

Then, instead of using `from nltk.tokenize import _get_punkt_tokenizer` we can directly use `from nltk.tokenize.punkt import PunktTokenizer` ([ref](https://github.com/nltk/nltk/blob/7397ccfed06e7c836d3acb0b9197f6e6b26c6741/nltk/tokenize/__init__.py#L96-L105)).

Note that `PunktTokenizer` is a subclass of `TokenizerI`:
```python
from nltk.tokenize.punkt import PunktTokenizer
from nltk.tokenize.api import TokenizerI

issubclass(PunktTokenizer, TokenizerI)
True
```


However, the `_get_punkt_tokenizer` is decorated with `@functools.lru_cache` that aims to utilize a lru cache for performance... then if we just do these changes, we should:
- use the `@functools.lru_cache` directly in `langchain` (e.g. [ref](https://github.com/langchain-ai/langchain/blob/master/libs/core/langchain_core/language_models/llms.py#L67))
- otherwise, just don't use it... but in this case, I'll run some benchmarking to understand the possible drawbacks